torch/_functorch/pytree_hacks.py

```diff
@@ -4,21 +4,8 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
-from torch.utils._pytree import tree_flatten, tree_unflatten
+# TODO: remove this file when the migration of the pytree utility is done
```

Comment:
@XuehaiPan, I want to figure out what exactly our plan is. For example, is optree going to stay out-of-tree, or do we want to merge it into PyTorch at some point?

IMO, we should merge optree into PyTorch. People are going to add new PyTree APIs, and to do so, they need to modify optree. optree being out-of-tree makes this a large maintainence burden for us.

A second point is: we can never delete torch._utils.pytree. In the long-term, we need a C++ pytree, but we also need a reference Python version of C++ pytree. The way Dynamo support for pytree is implemented (cc @ysiraichi) is that we trace through the reference python version of pytree.

Reply:
Thanks for your thoughts @zou3519!

> IMO, we should merge optree into PyTorch. People are going to add new PyTree APIs, and to do so, they need to modify optree. optree being out-of-tree makes this a large maintainence burden for us.

I understand and agree with that. Here are some opinions from my point of view. Correct me if I am wrong.

1. Currently, the APIs provided by `optree` can exactly simulate the behavior of `torch.utils._pytree` PyTree APIs (including functions and types, e.g., `TreeSpec`). `optree` provides low-level Python-based APIs and we can simulate exactly the C++ version via those functions. Here is an example:

    https://github.com/metaopt/optree/blob/72d9bf4e254966c8c319e854e8454e2d411bd767/tests/test_ops.py#L965-L1017

    If there is any concern about this, I am ready to submit a PR demonstrating how it works as expected.

2. Introducing new features into `torch.utils.pytree` is straightforward. It can be done by similar methods of adding extra registries for custom node types without modifying `optree`'s source code. Here are examples from JAX's PyTree:

    - [`jax.tree_utils.register_pytree_with_keys`](https://github.com/google/jax/blob/d4965cd33549ebe66cd474617366aa59b894d676/jax/_src/tree_util.py#L667-L707): add extra arguments to the register function.
    - [`jax.tree_utils.tree_flatten_with_path`](https://github.com/google/jax/blob/d4965cd33549ebe66cd474617366aa59b894d676/jax/_src/tree_util.py#L764-L778): flatten using the Python-based registry cooperating with the C++ flatten function.

    Note that the newest version of JAX's PyTree C extension does not support tree operations with paths. It is using its Python APIs as I mentioned above. (In `optree`, we implement tree operations with paths in C++ and they are much faster)

    When we need to add new features to `torch.utils.pytree`, we can do something similar. E.g., add an extra registry for custom node types without touching the C++ code.

    ```python
    __SERIALIZATION_REGISTRY = {
        list: ...,
        tuple: ...,
        dict: ...,
        ...
    }

    __PPRINT_REGISTRY = {
        list: ...,
        tuple: ...,
        dict: ...,
        ...
    }

    def register_pytree_node(
        cls,
        flatten_func,
        unflatten_func,
        *,
        to_serializable_func=None,    # for custom serialization
        from_serializable_func=None,
        pprint_func=None,             # for custom pprint
    ):
        optree.register_pytree_node(cls, flatten_func, unflatten_func, namespace='torch')
        __SERIALIZATION_REGISTRY[cls] = (to_serializable_func, from_serializable_func)
        __PPRINT_REGISTRY[cls] = pprint_func

    def treespec_dumps(treespec):
        if treespec.type in __SERIALIZATION_REGISTRY:
            return (to_serializable_func(treespec),) + tuple(treespec_dumps(child) for child in treespec.children())
        if optree.is_namedtuple(treespec.type):
            return ...  # case for namedtuple
        if optree.is_structseq(treespec.type):
            return ...  # case for PyStructSequence
        if treespec.type is None:
            return ...  # case for leaf
        raise NotImplementedError

    def treespec_loads(serialized):
        ...  # use __SERIALIZATION_REGISTRY

    def treespec_pprint(treespec):
        ...  # use __PPRINT_REGISTRY
    ```

3. Our intention with `optree` is to maintain it as a completely standalone library that can be used across different projects. This aligns with the initial need for PyTree implementation in PyTorch and our project ([TorchOpt](https://github.com/metaopt/torchopt.git)). This is also the motivation that we add the `namespace` isolation feature for the PyTree registry. We've found that many other libraries might have similar requirements, but are hesitant to install heavy dependencies such as JAX or PyTorch just to get PyTree support.

4. We (`optree` maintainers) will continue to independently maintain `optree` and are open to the idea of adding PyTorch team members in this repository as maintainers, or even admins. If you deem it more suitable, the PyTorch team can also create a fork of the library and maintain it independently in accordance with PyTorch-specific requirements.

We are open to discussing any other solutions and ideas to improve the PyTree APIs and the PyTorch ecosystem. Let's join hands to improve the open-source ecosystem around PyTree!

------

About the modification of `torch._functorch.pytree_hacks`: I moved those functions to `torch.utils._pytree` and updated the import statements.

Reply:
Thanks for the reply. Here's some thoughts and responses:

> We (optree maintainers) will continue to independently maintain optree and are open to the idea of adding PyTorch team members in this repository as maintainers, or even admins. If you deem it more suitable, the PyTorch team can also create a fork of the library and maintain it independently in accordance with PyTorch-specific requirements.

For now, I think we can keep on going with the plan of "have optree be an optional dependency". In the medium-term future we may want it to be a required dependency so that there isn't a large performance cliff when someone doesn't use optree. When that happens, it's possible that we'll want to fork optree, but ideally we would avoid that as much as possible and explore other solutions. Throwing some quick ideas out there: (1) we could take optree as a submodule, but then we would want to reduce the build dependencies (e.g. abseil); (2) if optree were header-only, then that could make distribution easier.

Reply:
> (1) we could take optree as a submodule, but then we would want to reduce the build dependencies (e.g. abseil);

We are planning to remove the abseil-cpp dependency and go with the STL library instead. This should help us support more platforms. Right now we support Windows amd64/arm64, macOS amd64/arm64/universal2, and Linux amd64/aarch64. Also, we found that abseil-cpp poses some packaging challenges when we need to support various C++ compilers and their standard libraries (especially for conda-forge packages).

With this change, the only build dependencies for `optree` will be the Python includes and pybind11 headers, which are already the dependencies of PyTorch. We have one worry here. We might want to use the [`std::format`](https://en.cppreference.com/w/cpp/utility/format/format utility) to switch out the [`absl::strings` library](https://abseil.io/docs/cpp/guides/strings). But `std::format` needs the C++-20 standard which might be too new for now.

Reply:
> With this change, the only build dependencies for optree will be the Python includes and pybind11 headers, which are already the dependencies of PyTorch. 

That's awesome! 

> But std::format needs the C++-20 standard which might be too new for now.

Yeah, that's probably too new. Is it possible to backport it somehow?

Reply:
> Is it possible to backport it somehow?

We could refactor the string formatting part by employing `sprintf` or `std::stringstream`. In doing so, it may become verbose and potentially complex.

Reply:
> We are planning to remove the abseil-cpp dependency and go with the STL library instead.

Done by:

- metaopt/optree#84
- metaopt/optree#85

Reply:
@XuehaiPan fmtlib is the inspiration for std::format and we already have it as a dependency in PyTorch if you want to use that: https://github.com/fmtlib/fmt . It also does not require C++20, but can use C++20 features if available.

Reply:
@Skylion007 Thanks for the suggestion. We have already refactored with `std::ostringstream` for all complex string formatting and concatenation. Now we have `pybind11` as our only third-party build dependency.